### ISSUE TITLE
Automated template instantiation of pairs of type (Number, width) as needed in VectorizedArray

### ISSUE THREAD

**Issue by peterrum** [dealii/dealii#8364]
In PR #8342, I have introduced a second template parameter to `VectorizedArray<Number, width>`, such that the user can specify the length of the array. The width cannot be selected arbitraily, but is fixed for the given processor architecture (`__SSE2__`, `__AVX__`, `__AVX512F__`) and the optimization level (`DEAL_II_COMPILER_VECTORIZATION_LEVEL`). For example for Skylake (AVX-512 and maximal optimization level), following instantiations of `VectorizedArray<Number, width>` are possible (ignoring all other other types since we do not support for them vectorization):
```cpp
VectorizedArray<double, 1>
VectorizedArray<double, 2>
VectorizedArray<double, 4>
VectorizedArray<double, 8>
VectorizedArray<float, 1>
VectorizedArray<float, 4>
VectorizedArray<float, 8>
VectorizedArray<float, 16>
```
or expressed as a set of pairs (Number, width) \in {(double, 1), (double, 2), (double, 4), (double, 8), (float, 1), (float, 4), (float, 8), (float, 16)}.

In PR  #8348, I have added `VectorizedArrayType` to the `MatrixFree`-classes as an additionally template parameter to be able to control the vectorization strategies used. Currently, `VectorizedArrayType` is assuming that `VectorizedArray<Number, width>` is passed (statically asserted).

During this PR, I faced the problem that many functions/classes from the `MatrixFree`-framework (which now have an additional template parameter `VectorizedArrayType`) are explicitly instantiated in the `.inst.in`-files.     As a quick/dirty solution, I have manually instantiated all possible `VectorizedArrayType`s (8!) with a lot of ifdefs(!). As an example, see https://github.com/dealii/dealii/blob/7a2ab016ee28009d4be097f21bb0bbf8074915ab/source/matrix_free/mapping_info.inst.in. You can find similar copy-and-pastes in the other `.inst.in`-files in the matrixfree folder. This is not a solution...

What we want: is to loop over all possible pairs of `Number` and `width` (collected in `REAL_SCALARS_V`) in the following way during instantiations:
```cpp
for (deal_II_dimension : DIMENSIONS; deal_II_scalar_v : REAL_SCALARS_V)
  {
    template struct internal::MatrixFreeFunctions::MappingInfo<
      deal_II_dimension,
      deal_II_scalar_v::value_type,
      VectorizedArray<deal_II_scalar_v::value_type,
                      deal_II_scalar_v::n_array_elements>>;
  }
```
Please note the `deal_II_scalar_v::value_type` and `deal_II_scalar_v::n_array_elements` where I access the first and second entry of the tuple.

The problem in this regard is that the instantiation script does - as far as I can tell - not support tuples (note: that it is not a solution to regard one tuple as one string, since we explicitly need to access both the value type and the width!). 

As a remedy, there might be two possible solutions:
- Add to `cmake/config/template-arguments.in` lists of tuple like structs:
```cpp
REAL_SCALARS_V    := { VectorizedArray<1, double>; VectorizedArray<1, float>; ...}
```
- Extend the instantiation script with tuple capabilities. With the capability that the user can access the first and the second entry separately. 

I have implemented in collaboration with @kronbichler  the first version on https://github.com/peterrum/dealii/commit/f33412f137e98d37980df8c05deb893c2ab60309 and integrated exemplarily in mapping_info.inst.in. This approach seems to be straightforward. The second version might be worth implementing if we see other occurrences where we might use pairs during instantiation. Are there any?

Since this is the first time I have to mess with the issue instantiation (and I might have overlooked something, and/or my solution does fit into the current philosophy of instantiation, and/or there is a much better solution I was not thinking of) I would like to discuss the presented issue first. I will open a new PR as soon as we found a solution we all are satisfied with.

**Comment by peterrum** [comment]
I just noticed that I can simplify the above code:
```cpp
for (deal_II_dimension : DIMENSIONS; deal_II_scalar_v : REAL_SCALARS_V)
  {
    template struct internal::MatrixFreeFunctions::MappingInfo<
      deal_II_dimension,
      deal_II_scalar_v::value_type, 
      deal_II_scalar_v>;
  }
```

**Comment by masterleinad** [comment]
That looks pretty reasonable to me.

**Comment by peterrum** [comment]
@masterleinad: thanks for the feedback!

**Comment by bangerth** [comment]
I think you can do this simpler: let's say you've defined in `template-arguments.in` a list `vector_lengths`, then you can do the following in the `.inst.in` files:
```
  for (vector_length : vector_lengths)
  {
#if vector_length < 16
    ... instantiate the double and float types...
#else
  // vector length is 16, which also supports floats
   ... instantiate the float types only...
#endif
  }
```
This way you won't have to extend the `expand_instantiations` script.

**Comment by peterrum** [comment]
@bangerth: Thanks for your feedback. I am not sure what are referring to? If we write in a style as you proposed, we would have to add an additional ifdef:
```cpp
  for (vector_length : vector_lengths)
  {
#if vector_length < 16
  #if vector_length == 2
     //... instantiate the double types only...
  #else
     // ... instantiate the double and float types...
  #endif
#else
  // ... instantiate the float types only...
#endif
  }
```
This would lead to the fact that we have to write practically the same code three times.

**Comment by bangerth** [comment]
Yes, something like this. I only wanted to point out that it's not necessary to extend `expand_instantiations.cc` or use non-standard pairs of values into the `template-arguments` file.  In other words, it is not necessary to create a *general* solution to address a *single occurrence*.

But it may be that you've already found another way to solve this, and in that case you should feel free to completely ignore my comment :-)

**Comment by peterrum** [comment]
@bangerth: I just noticed this is the issue. The related PR is #8383. I think the solution described there is - as it is now - quite nice. If you want you can have a look at it. Hope you approve it ;)

**Comment by tamiko** [comment]
I'll close this issue - it looks like the p :-)

**Comment by bangerth** [comment]
Yes, I agree that the way you solved this in #8383 seems like the right way to go.
